### PR TITLE
Drop the use of the deprecated setuptools Features

### PR DESCRIPTION
**Pull request by jamadden**
https://github.com/pypa/setuptools/issues/65

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -22,39 +22,37 @@
...
 extra = {
-    'extras_require': {
-        'testing': ['nose', 'coverage'],
-        'docs': ['Sphinx'],
-    },
+
 }
 
+ext_modules = []
 if not is_pypy and not is_jython:
```
Comment (mgedmin) [inline setup.py:48]: Maybe please add an `and not is_pure` check with the os.getenv('PURE_PYTHON') logic copied from zope.interface etc.? So people can ask to skip extension modules.

File: setup.py [hunk 2]
```diff
@@ -141,5 +133,11 @@ def _unavailable(self, e):
     test_suite='zope.i18nmessageid.tests.test_suite',
     zip_safe=False,
     cmdclass={'build_ext': optional_build_ext},
-    **extra
+    ext_modules=ext_modules,
+    extras_require={
+        'testing': [
+            'nose', 'coverage'
```
Comment (mgedmin) [inline setup.py]: Do we actually allow `nose` as a test runner here?  AFAIU nose is deprecated upstream.

File: setup.py [hunk 3]
```diff
@@ -69,20 +67,14 @@ class optional_build_ext(build_ext):
...
     def build_extension(self, ext):
         try:
             build_ext.build_extension(self, ext)
-
-        except (CCompilerError, DistutilsExecError):
-            # The sys.exc_info()[1] is to preserve compatibility with both
-            # Python 2.5 and 3.x, which is needed in setup.py.
-            self._unavailable(sys.exc_info()[1])
+        except (CCompilerError, DistutilsExecError) as e:
+            self._unavailable(e)
```
Comment (mgedmin) [inline setup.py:79]: :+1:

File: tox.ini [hunk 4]
```diff
@@ -5,7 +5,6 @@ envlist =
 [testenv]
 deps =
      .[test]
-     zope.testrunner
```
Comment (mgedmin) [inline tox.ini:8]: `extras = test` would be better than `deps = .[test]`

**Review by mgedmin — APPROVED** [review]
LGTM!

File: tox.ini (hunk 4 above)
Comment (jamadden) [reply tox.ini:8]: I knew there was another way to spell that but I always forget what it is.

File: setup.py (hunk 1 above)
Comment (mgedmin) [reply setup.py:48]: You know what, you're right and I was wrong.  Build-time checking of PURE_PYTHON will poison pip's wheel cache and we should fix zope.interface not to do that.

Comment (jamadden) [reply setup.py:48]: I was slightly surprised to discover that you get different wheel file names depending on whether PURE_PYTHON is set (`zope.interface-4.7.2.dev0-py3-none-any.whl`) or not (`zope.interface-4.7.2.dev0-cp38-cp38-macosx_10_15_x86_64.whl`). I suspect it still messes with the cache though.

**Pull request merged**